### PR TITLE
python3Packages.cffi: fix tests on FreeBSD

### PR DESCRIPTION
**Pull request by rhelmot**
## Description of changes

Followup to https://github.com/NixOS/nixpkgs/pull/328776, depends on https://github.com/NixOS/nixpkgs/pull/328943. See both for context.

I don't believe either of these changes are worth submitting upstream. One is modifying the existing patch to do what it claims to do and not just scoped to darwin, and one is just disabling a test. It's kind of a miracle that this is all it takes to pass tests, seeing as there are various comments to the effect of assuming that "anything that's not darwin or win32 must be linux and act exactly like linux". FreeBSD's libdl (the subject of the skipped test) claims to support dlopen, but it [actually only has a stub](https://github.com/freebsd/freebsd-src/blob/559a218c9b257775fb249b67945fe4a05b7a6b9f/lib/libc/gen/dlfcn.c#L111-L118), and if you want the real dlopen you have to link against libc, but you actually need the symbol from the dynamic loader? It's a bit of a mess and I want out of here.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Ericson2314 — COMMENTED** [review]
The first part, about the clang issue, should go in the commit message.

The second part, about the disabled test, should be a code comment.

**Review by mweinelt — CHANGES_REQUESTED** [review]
Please send this upstream.

**Comment by Ericson2314** [comment]
> I don't believe either of these changes are worth submitting upstream. 

What @rhelmot wrote on this subject. Keep in mind we are modifying an existing not-upstreamed patch?

**Comment by Ericson2314** [comment]
There is still a PR description that should be a commit message, however.

**Comment by rhelmot** [comment]
omg @Ericson2314 I am so sorry, I rebased the old version of the patch before I addressed your original review comments. They should be fixed now.

**Comment by mweinelt** [comment]
@reckenrode @tjni @abathur Please upstream the patches that we currently have on cffi.

I'm not at all happy to keep them for years and maintain them downstream.

**Comment by reckenrode** [comment]
`darwin-use-libffi-closures.diff` is a workaround for the old SDK on x86_64-darwin. It can be dropped for 25.05 since the default will be newer than 10.15.

`clang-pointer-substraction-warning.diff` can probably be dropped. Upstream addressed the issue in https://github.com/python-cffi/cffi/pull/78.

**Comment by mweinelt** [comment]
Okay, then we are back to reporting this issue and a potential patch upstream, before we choose to pick it downstream.

**Comment by emilazy** [comment]
The patch isn’t actually specific to old SDK versions, it’s specific to using upstream `libffi` on macOS rather than Apple’s fork. Options include depending on `darwin.libffi` to just use the real thing, setting the `-DFFI_AVAILABLE_APPLE` compile flag, or just removing the condition upstream since https://groups.google.com/g/python-cffi/c/xU0Usa8dvhk seems lukewarm on that.

**Comment by emilazy** [comment]
FWIW the `postPatch` stuff is obsolete.

File: pkgs/development/python-modules/cffi/default.nix [hunk 1]
```diff
@@ -57,6 +48,11 @@ else
         --replace '/usr/include/libffi' '${lib.getDev libffi}/include'
     '';
 
+    disabledTests = lib.optionals stdenv.isFreeBSD [
```
Comment (mweinelt) [inline pkgs/development/python-modules/cffi/default.nix:51]:
```suggestion
    disabledTests = lib.optionals stdenv.hostPlatform.isFreeBSD [
```

see https://github.com/NixOS/nixpkgs/pull/341407

**Review by mweinelt — APPROVED** [review]

**Comment by mweinelt** [comment]
Eval failures are unrelated.

**Pull request merged**